File: packages/dfos-web-relay/src/bootstrap.ts
/*

  BOOTSTRAP RELAY IDENTITY

  Generate a JIT (just-in-time) relay identity and profile artifact. Used when
  createRelay is called without a pre-created identity.

*/

import {
  encodeEd25519Multikey,
  signArtifact,
  signIdentityOperation,
  type ArtifactPayload,
  type IdentityOperation,
} from '@metalabel/dfos-protocol/chain';
import {
  createNewEd25519Keypair,
  generateId,
  importEd25519Keypair,
  signPayloadEd25519,
} from '@metalabel/dfos-protocol/crypto';
import { ingestOperations } from './ingest';
import type { RelayIdentity, RelayStore } from './types';

/**
 * Generate a relay identity and profile artifact, ingest both into the store
 *
 * Creates an Ed25519 keypair, signs an identity genesis operation, derives
 * the DID, then signs a profile artifact with the relay's name. Both the
 * identity genesis and profile artifact are ingested into the store so
 * they are available via the relay's proof plane routes.
 */
export const bootstrapRelayIdentity = async (store: RelayStore): Promise<RelayIdentity> => {
  const keypair = createNewEd25519Keypair();
  const keyId = generateId('key');
  return bootstrapWithKeyMaterial(store, {
    privateKey: keypair.privateKey,
    publicKey: keypair.publicKey,
    keyId,
  });
};

/**
 * Bootstrap a relay identity from an EXISTING key + key ID, with optional pinned
 * timestamps and profile name. Used for deterministic bootstrap — e.g. the
 * dual-relay parity harness pins one key + one createdAt across both twins so
 * the relay's own genesis + profile log entries are byte-identical, and durable
 * relays that reload their key from storage. Mirrors the Go twin's
 * BootstrapRelayIdentityFromKey.
 */
export const bootstrapRelayIdentityFromKey = async (
  store: RelayStore,
  params: {
    privateKey: Uint8Array;
    keyId: string;
    name?: string;
    createdAt?: string;
  },
): Promise<RelayIdentity> => {
  const { publicKey } = importEd25519Keypair(params.privateKey);
  return bootstrapWithKeyMaterial(store, {
    privateKey: params.privateKey,
    publicKey,
    keyId: params.keyId,
    ...(params.name !== undefined ? { name: params.name } : {}),
    ...(params.createdAt !== undefined ? { createdAt: params.createdAt } : {}),
  });
};

/**
 * Shared bootstrap core: sign + ingest the identity genesis and profile
 * artifact for the given key material. The JIT path passes a random key and
 * lets createdAt default to now; the from-key path pins both.
 */
const bootstrapWithKeyMaterial = async (
  store: RelayStore,
  params: {
    privateKey: Uint8Array;
    publicKey: Uint8Array;
    keyId: string;
    name?: string;
    createdAt?: string;
  },
): Promise<RelayIdentity> => {
  const { privateKey, publicKey, keyId } = params;
  const name = params.name ?? 'DFOS Relay';
  const createdAt = params.createdAt ?? new Date().toISOString();
  const multibase = encodeEd25519Multikey(publicKey);
  const signer = async (msg: Uint8Array) => signPayloadEd25519(msg, privateKey);

  const key = { id: keyId, type: 'Multikey' as const, publicKeyMultibase: multibase };

  // --- identity genesis ---

  const identityOp: IdentityOperation = {
    version: 1,
    type: 'create',
    authKeys: [key],
    assertKeys: [key],
    controllerKeys: [key],
    createdAt,
  };

  const { jwsToken: identityJws } = await signIdentityOperation({
    operation: identityOp,
    signer,
    keyId,
  });

  // ingest identity genesis to derive the DID
  const [identityResult] = await ingestOperations([identityJws], store);
  if (!identityResult || identityResult.status === 'rejected' || !identityResult.chainId) {
    throw new Error(`failed to bootstrap relay identity: ${identityResult?.error ?? 'unknown'}`);
  }
  const did = identityResult.chainId;

  // --- profile artifact ---

  const profilePayload: ArtifactPayload = {
    version: 1,
    type: 'artifact',
    did,
    content: {
      $schema: 'https://schemas.dfos.com/profile/v1',
      name,
    },
    createdAt,
  };

  const kid = `${did}#${keyId}`;
  const { jwsToken: profileArtifactJws } = await signArtifact({
    payload: profilePayload,
    signer,
    kid,
  });

  // ingest profile artifact into the store
  const [artifactResult] = await ingestOperations([profileArtifactJws], store);
  if (!artifactResult || artifactResult.status === 'rejected') {
    throw new Error(
      `failed to ingest relay profile artifact: ${artifactResult?.error ?? 'unknown'}`,
    );
  }

  return { did, profileArtifactJws };
};

File: packages/dfos-web-relay-go/bootstrap.go
package relay

import (
	"crypto/ed25519"
	"crypto/rand"
	"fmt"

	dfos "github.com/metalabel/dfos/packages/dfos-protocol-go"
)

// ProfileConfig holds the relay's profile fields. If empty, defaults are used.
type ProfileConfig struct {
	Name        string
	Description string
}

// BootstrapRelayIdentity generates a JIT relay identity and profile artifact,
// ingests both into the store, and returns the relay identity.
func BootstrapRelayIdentity(store Store) (*RelayIdentity, error) {
	return bootstrapRelay(store, nil, nil, ProfileConfig{})
}

// BootstrapRelayIdentityWithProfile generates a JIT relay identity with a
// custom profile, ingests both into the store, and returns the relay identity.
func BootstrapRelayIdentityWithProfile(store Store, profile ProfileConfig) (*RelayIdentity, error) {
	return bootstrapRelay(store, nil, nil, profile)
}

// BootstrapRelayIdentityFromKey creates a relay identity from an existing
// private key and key ID, signs a profile artifact, ingests both, and returns
// the relay identity. Used for persistent bootstrap where keys are loaded from
// storage.
func BootstrapRelayIdentityFromKey(store Store, priv ed25519.PrivateKey, keyID string, profile ProfileConfig) (*RelayIdentity, error) {
	pub := priv.Public().(ed25519.PublicKey)
	return bootstrapRelay(store, priv, &keyMaterial{pub: pub, keyID: keyID}, profile)
}

// RebootstrapProfile signs a fresh profile artifact for an existing relay
// identity. The identity chain is already in the store — this only produces a
// new profile artifact (e.g. after RELAY_NAME changes between restarts).
func RebootstrapProfile(store Store, priv ed25519.PrivateKey, keyID, did string, profile ProfileConfig) (*RelayIdentity, error) {
	kid := did + "#" + keyID
	content := map[string]any{
		"$schema": "https://schemas.dfos.com/profile/v1",
		"name":    profile.Name,
	}
	if content["name"] == "" {
		content["name"] = "DFOS Relay"
	}
	if profile.Description != "" {
		content["description"] = profile.Description
	}

	profileArtifactJWS, _, err := dfos.SignArtifact(did, content, kid, priv)
	if err != nil {
		return nil, fmt.Errorf("sign profile artifact: %w", err)
	}

	// ingest profile artifact (new artifact each boot — immutable, append-only)
	artResults := IngestOperations([]string{profileArtifactJWS}, store)
	if len(artResults) == 0 || artResults[0].Status == "rejected" {
		errMsg := "unknown"
		if len(artResults) > 0 && artResults[0].Error != "" {
			errMsg = artResults[0].Error
		}
		return nil, fmt.Errorf("ingest relay profile artifact: %s", errMsg)
	}

	return &RelayIdentity{
		DID:                did,
		ProfileArtifactJWS: profileArtifactJWS,
		PrivateKey:         priv,
		KeyID:              keyID,
	}, nil
}

type keyMaterial struct {
	pub   ed25519.PublicKey
	keyID string
}

func bootstrapRelay(store Store, priv ed25519.PrivateKey, existing *keyMaterial, profile ProfileConfig) (*RelayIdentity, error) {
	var pub ed25519.PublicKey
	var keyID string
	var err error

	if existing != nil {
		pub = existing.pub
		keyID = existing.keyID
	} else {
		pub, priv, err = ed25519.GenerateKey(rand.Reader)
		if err != nil {
			return nil, fmt.Errorf("generate keypair: %w", err)
		}
		keyID = dfos.GenerateKeyID()
	}

	mk := dfos.NewMultikeyPublicKey(keyID, pub)

	// sign identity genesis
	identityJWS, did, _, err := dfos.SignIdentityCreate(
		[]dfos.MultikeyPublicKey{mk}, // controller
		[]dfos.MultikeyPublicKey{mk}, // auth
		[]dfos.MultikeyPublicKey{mk}, // assert
		keyID,
		priv,
	)
	if err != nil {
		return nil, fmt.Errorf("sign identity genesis: %w", err)
	}

	// ingest identity genesis — may return duplicate if key was reloaded
	results := IngestOperations([]string{identityJWS}, store)
	if len(results) == 0 || results[0].Status == "rejected" {
		errMsg := "unknown"
		if len(results) > 0 && results[0].Error != "" {
			errMsg = results[0].Error
		}
		return nil, fmt.Errorf("bootstrap relay identity: %s", errMsg)
	}

	// sign profile artifact
	kid := did + "#" + keyID
	content := map[string]any{
		"$schema": "https://schemas.dfos.com/profile/v1",
		"name":    profile.Name,
	}
	if content["name"] == "" {
		content["name"] = "DFOS Relay"
	}
	if profile.Description != "" {
		content["description"] = profile.Description
	}

	profileArtifactJWS, _, err := dfos.SignArtifact(did, content, kid, priv)
	if err != nil {
		return nil, fmt.Errorf("sign profile artifact: %w", err)
	}

	// ingest profile artifact
	artResults := IngestOperations([]string{profileArtifactJWS}, store)
	if len(artResults) == 0 || artResults[0].Status == "rejected" {
		errMsg := "unknown"
		if len(artResults) > 0 && artResults[0].Error != "" {
			errMsg = artResults[0].Error
		}
		return nil, fmt.Errorf("ingest relay profile artifact: %s", errMsg)
	}

	return &RelayIdentity{
		DID:                did,
		ProfileArtifactJWS: profileArtifactJWS,
		PrivateKey:         priv,
		KeyID:              keyID,
	}, nil
}
